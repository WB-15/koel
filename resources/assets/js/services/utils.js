export default {
    /**
     * Convert a duration in seconds into H:i:s format.
     * If H is 0, it will be ommited.
     */
    secondsToHis(d) {
        d = parseInt(d);
        
        var s = d%60;

        if (s < 10) {
            s = '0' + s;
        }

        var i = Math.floor((d/60)%60);

        if (i < 10) {
            i = '0' + i;
        }

        var h = Math.floor(d/3600);

        if (h < 10) {
            h = '0' + h;
        }

        return (h === '00' ? '' : h + ':') + i + ':' + s;
    },

    /**
     * Parse the validation error from the server into a flattened array of messages.
     * 
     * @param  {Object}  error  The error object in JSON format.
     *                                    
     * @return {Array}
     */
    parseValidationError(error) {
        return Object.keys(error).reduce((messages, field) => messages.concat(error[field]), []);
    }
};
