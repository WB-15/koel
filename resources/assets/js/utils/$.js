export const $ = {
  is (el, selector) {
    return (el.matches ||
      el.matchesSelector ||
      el.msMatchesSelector ||
      el.mozMatchesSelector ||
      el.webkitMatchesSelector ||
      el.oMatchesSelector).call(el, selector)
  },

  addClass (el, className) {
    if (!el) {
      return
    }

    if (el.classList) {
      el.classList.add(className)
    } else {
      el.className += ` ${className}`
    }
  },

  removeClass (el, className) {
    if (!el) {
      return
    }

    if (el.classList) {
      el.classList.remove(className)
    } else {
      el.className = el.className.replace(new RegExp('(^|\\b)' + className.split(' ').join('|') + '(\\b|$)', 'gi'), ' ')
    }
  },

  scrollTo (el, to, duration) {
    if (duration <= 0 || !el) {
      return
    }
    const difference = to - el.scrollTop
    const perTick = difference / duration * 10
    window.setTimeout(() => {
      el.scrollTop = el.scrollTop + perTick
      if (el.scrollTop === to) {
        return
      }
      this.scrollTo(el, to, duration - 10);
    }, 10)
  }
}
