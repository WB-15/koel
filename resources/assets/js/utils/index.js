export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
