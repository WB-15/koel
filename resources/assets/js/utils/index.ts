export * from './event'
export * from './alerts'
export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './helpers'
export * from './fileReader'
export * from './directoryReader'
