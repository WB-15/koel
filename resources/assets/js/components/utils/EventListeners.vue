<template>
  <span></span>
</template>

<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { authService } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.is_smart && !playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    authService.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
