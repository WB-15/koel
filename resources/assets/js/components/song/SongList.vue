<template>
  <div
    class="song-list-wrap main-scroll-wrap"
    :class="type"
    ref="wrapper"
    tabindex="0"
    @keydown.delete.prevent.stop="handleDelete"
    @keydown.enter.prevent.stop="handleEnter"
    @keydown.a.prevent="handleA"
  >
    <div class="song-list-header" :class="mergedConfig.sortable ? 'sortable' : 'unsortable'">
      <span @click="sort('song.track')" class="track-number" v-if="mergedConfig.columns.includes('track')">
        #
        <i class="fa fa-angle-down" v-show="primarySortField === 'song.track' && sortOrder === 'Asc'"></i>
        <i class="fa fa-angle-up" v-show="primarySortField === 'song.track' && sortOrder === 'Desc'"></i>
      </span>
      <span @click="sort('song.title')" class="title" v-if="mergedConfig.columns.includes('title')">
        Title
        <i class="fa fa-angle-down" v-show="primarySortField === 'song.title' && sortOrder === 'Asc'"></i>
        <i class="fa fa-angle-up" v-show="primarySortField === 'song.title' && sortOrder === 'Desc'"></i>
      </span>
      <span
        @click="sort(['song.album.artist.name', 'song.album.name', 'song.track'])"
        class="artist"
        v-if="mergedConfig.columns.includes('artist')"
      >
        Artist
        <i class="fa fa-angle-down" v-show="primarySortField === 'song.album.artist.name' && sortOrder === 'Asc'"></i>
        <i class="fa fa-angle-up" v-show="primarySortField === 'song.album.artist.name' && sortOrder === 'Desc'"></i>
      </span>
      <span
        @click="sort(['song.album.name', 'song.track'])"
        class="album"
        v-if="mergedConfig.columns.includes('album')"
      >
        Album
        <i class="fa fa-angle-down" v-show="primarySortField === 'song.album.name' && sortOrder === 'Asc'"></i>
        <i class="fa fa-angle-up" v-show="primarySortField === 'song.album.name' && sortOrder === 'Desc'"></i>
      </span>
      <span @click="sort('song.length')" class="time" v-if="mergedConfig.columns.includes('length')">
        <i class="fa fa-angle-down" v-show="primarySortField === 'song.length' && sortOrder === 'Asc'"></i>
        <i class="fa fa-angle-up" v-show="primarySortField === 'song.length' && sortOrder === 'Desc'"></i>
        &nbsp;<i class="duration-header fa fa-clock-o"></i>
      </span>
      <span class="favorite"></span>
      <span class="play"></span>
    </div>

    <VirtualScroller v-slot="{ item }" :item-height="35" :items="songProxies">
      <SongListItem :item="item" :columns="mergedConfig.columns" :key="item.song.id"/>
    </VirtualScroller>
  </div>
</template>

<script lang="ts">
export default {
  name: 'SongList'
}
</script>

<script lang="ts" setup>
import isMobile from 'ismobilejs'

import {
  computed,
  defineAsyncComponent,
  getCurrentInstance,
  nextTick,
  onMounted,
  ref,
  toRefs,
  watch
} from 'vue'

import { $, eventBus, orderBy, startDragging, arrayify } from '@/utils'
import { queueStore } from '@/stores'

type SortField = 'song.track'
  | 'song.disc'
  | 'song.title'
  | 'song.album.artist.name'
  | 'song.album.name'
  | 'song.length'

type SortOrder = 'Asc' | 'Desc' | 'None'

interface SongRow {
  props: {
    item: SongProxy
  }
}

const VirtualScroller = defineAsyncComponent(() => import('@/components/ui/VirtualScroller.vue'))
const SongListItem = defineAsyncComponent(() => import('@/components/song/SongListItem.vue'))

const props = withDefaults(
  defineProps<{ items: Song[], type?: SongListType, config?: Partial<SongListConfig> }>(),
  { type: 'all-songs', config: () => ({}) }
)

const { items, type, config } = toRefs(props)

const lastSelectedRow = ref<SongRow>()
const sortFields = ref<SortField[]>([])
const sortOrder = ref<SortOrder>('None')
const songProxies = ref<SongProxy[]>([])

const allowSongReordering = computed(() => type.value === 'queue')
const selectedSongs = computed(() => songProxies.value.filter(row => row.selected).map(row => row.song))
const primarySortField = computed(() => sortFields.value.length === 0 ? null : sortFields.value[0])

const mergedConfig = computed((): SongListConfig => {
  return Object.assign({
    sortable: true,
    columns: ['track', 'title', 'artist', 'album', 'length']
  }, config.value)
})

/**
 * Since song objects themselves are shared by all song lists, we can't use them directly to
 * determine their selection status (selected/unselected). Therefore, for each song list, we
 * maintain an array of "song proxies," each containing the song itself and the "selected" flag.
 */
const generateSongProxies = () => {
  // Since this method re-generates the song wrappers, we need to keep track of  the
  // selected songs manually.
  const selectedSongIds = selectedSongs.value.map(song => song.id)

  return items.value.map(song => ({
    song,
    selected: selectedSongIds.includes(song.id)
  }))
}

const nextSortOrder = computed<SortOrder>(() => {
  if (sortOrder.value === 'None') return 'Asc'
  if (sortOrder.value === 'Asc') return 'Desc'
  return 'None'
})

const sort = (field: SortField | SortField[] = [], order: SortOrder | null = null) => {
  // there are certain circumstances where sorting is simply disallowed, e.g. in Queue
  if (!mergedConfig.value.sortable) {
    return
  }

  sortFields.value = arrayify(field)

  if (!sortFields.value.length && ['album', 'artist'].includes(type.value)) {
    // by default, sort Album/Artist by track numbers for a more friendly UX
    sortFields.value.push('song.track')
    order = 'Asc'
  }

  if (sortFields.value.includes('song.track') && !sortFields.value.includes('song.disc')) {
    // Track numbers should always go in conjunction with disc numbers.
    sortFields.value.push('song.disc')
  }

  sortOrder.value = order === null ? nextSortOrder.value : order

  songProxies.value = sortOrder.value === 'None'
    ? generateSongProxies()
    : orderBy(songProxies.value, sortFields.value, sortOrder.value === 'Desc')
}

const render = () => {
  mergedConfig.value.sortable || (sortFields.value = [])
  songProxies.value = generateSongProxies()
  sort(sortFields.value, sortOrder.value)
}

watch(items, () => render())

const vm = getCurrentInstance()
watch(selectedSongs, () => eventBus.emit('SET_SELECTED_SONGS', selectedSongs.value, vm?.parent))

const emit = defineEmits(['press:enter', 'press:delete'])

const handleDelete = () => {
  emit('press:delete')
  clearSelection()
}

const handleEnter = (event: KeyboardEvent) => {
  emit('press:enter', event)
  clearSelection()
}

/**
 * Select all (filtered) rows in the current list.
 */
const selectAllRows = () => songProxies.value.forEach(row => (row.selected = true))
const clearSelection = () => songProxies.value.forEach(row => (row.selected = false))

const handleA = (event: KeyboardEvent) => (event.ctrlKey || event.metaKey) && selectAllRows()

const rowClicked = (rowVm: SongRow, event: MouseEvent) => {
  // If we're on a touch device, or if Ctrl/Cmd key is pressed, just toggle selection.
  if (isMobile.any) {
    toggleRow(rowVm)
    return
  }

  if (event.ctrlKey || event.metaKey) {
    toggleRow(rowVm)
  }

  if (event.button === 0) {
    if (!(event.ctrlKey || event.metaKey || event.shiftKey)) {
      clearSelection()
      toggleRow(rowVm)
    }

    if (event.shiftKey && lastSelectedRow.value) {
      selectRowsBetween(lastSelectedRow.value, rowVm)
    }
  }
}

const toggleRow = (rowVm: SongRow) => {
  rowVm.props.item.selected = !rowVm.props.item.selected
  lastSelectedRow.value = rowVm
}

const selectRowsBetween = (firstRowVm: SongRow, secondRowVm: SongRow) => {
  const indexes = [
    songProxies.value.indexOf(firstRowVm.props.item),
    songProxies.value.indexOf(secondRowVm.props.item)
  ]

  indexes.sort((a, b) => a - b)

  for (let i = indexes[0]; i <= indexes[1]; ++i) {
    songProxies.value[i].selected = true
  }
}

/**
 * Enable dragging songs by capturing the dragstart event on a table row.
 * Even though the event is triggered on one row only, we'll collect other
 * selected rows, if any, as well.
 */
const dragStart = (rowVm: SongRow, event: DragEvent) => {
  // If the user is dragging an unselected row, clear the current selection.
  if (!rowVm.props.item.selected) {
    clearSelection()
    rowVm.props.item.selected = true
  }

  startDragging(event, selectedSongs.value, 'Song')
}

/**
 * Add a "droppable" class and set the drop effect when other songs are dragged over a row.
 */
const allowDrop = (event: DragEvent) => {
  if (!allowSongReordering.value) {
    return
  }

  $.addClass((event.target as Element).parentElement, 'droppable')
  event.dataTransfer!.dropEffect = 'move'

  return false
}

/**
 * Perform reordering songs upon dropping if the current song list is of type Queue.
 */
const handleDrop = (rowVm: SongRow, event: DragEvent) => {
  if (
    !allowSongReordering.value ||
    !event.dataTransfer!.getData('application/x-koel.text+plain') ||
    !selectedSongs.value.length
  ) {
    return removeDroppableState(event)
  }

  queueStore.move(selectedSongs.value, rowVm.props.item.song)
  return removeDroppableState(event)
}

const removeDroppableState = (event: DragEvent) => {
  $.removeClass((event.target as Element).parentElement, 'droppable')
  return false
}

const openContextMenu = async (rowVm: SongRow, event: MouseEvent) => {
  // If the user is right-clicking an unselected row,
  // clear the current selection and select it instead.
  if (!rowVm.props.item.selected) {
    clearSelection()
    toggleRow(rowVm)
  }

  await nextTick()
  eventBus.emit('SONG_CONTEXT_MENU_REQUESTED', event, selectedSongs.value)
}

const getAllSongsWithSort = () => songProxies.value.map(proxy => proxy.song)

defineExpose({
  rowClicked,
  dragStart,
  allowDrop,
  handleDrop,
  removeDroppableState,
  openContextMenu,
  getAllSongsWithSort,
  sort
})

onMounted(() => render())
</script>

<style lang="scss">
.song-list-wrap {
  position: relative;
  padding: 0 !important;
  display: flex;
  flex-direction: column;

  .song-list-header {
    background: var(--color-bg-secondary);
    z-index: 1;
    display: flex;
  }

  div.droppable {
    border-bottom-width: 3px;
    border-bottom-color: var(--color-green);
  }

  .song-list-header span, .song-item span {
    text-align: left;
    padding: 8px;
    vertical-align: middle;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;

    &.time {
      flex-basis: 96px;
      padding-right: 24px;
      text-align: right;
    }

    &.track-number {
      flex-basis: 66px;
      padding-left: 24px;
    }

    &.artist {
      flex-basis: 23%;
    }

    &.album {
      flex-basis: 27%;
    }

    &.favorite {
      flex-basis: 36px;
    }

    &.play {
      display: none;

      @media (hover: none) {
        display: block;
        position: absolute;
        top: 8px;
        right: 4px;
      }
    }

    &.title {
      flex: 1;
    }
  }

  .song-list-header {
    color: var(--color-text-secondary);
    letter-spacing: 1px;
    text-transform: uppercase;
    cursor: pointer;

    i:not(.duration-header) {
      color: var(--color-highlight);
      font-size: 1.2rem;
    }
  }

  .unsortable span {
    cursor: default;
  }

  .scroller {
    overflow: auto;
    position: absolute;
    top: 35px;
    left: 0;
    bottom: 0;
    right: 0;
    overflow-y: scroll;
    -webkit-overflow-scrolling: touch;

    .item-container {
      position: absolute;
      left: 0;
      right: 0;
      min-height: 100%;
    }

    .item {
      margin-bottom: 0;
    }
  }

  @media only screen and (max-width: 768px) {
    .song-list-header {
      display: none;
    }

    .scroller {
      top: 0;

      .item-container {
        left: 12px;
        right: 12px;
        width: calc(100vw - 24px);
      }
    }

    .song-item {
      padding: 8px 32px 8px 4px;
      position: relative;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      color: var(--color-text-secondary);
      width: 100%;
    }

    .song-item span {
      display: none;
      padding: 0;
      vertical-align: bottom;
      color: var(--color-text-primary);

      &.artist, &.title {
        display: inline;
      }

      &.artist {
        color: var(--color-text-secondary);
        font-size: 0.9rem;
        padding: 0 4px;
      }
    }
  }
}
</style>
