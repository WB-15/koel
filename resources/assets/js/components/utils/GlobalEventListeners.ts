/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */

import { defineComponent } from 'vue'
import { authService } from '@/services'
import { playlistFolderStore, playlistStore, userStore } from '@/stores'
import { eventBus, forceReloadWindow, requireInjection } from '@/utils'
import { RouterKey } from '@/symbols'
import { useDialogBox, useMessageToaster } from '@/composables'

export const GlobalEventListeners = defineComponent({
  setup (props, { slots }) {
    const { toastSuccess } = useMessageToaster()
    const { showConfirmDialog } = useDialogBox()
    const router = requireInjection(RouterKey)

    eventBus.on('PLAYLIST_DELETE', async playlist => {
      if (await showConfirmDialog(`Delete the playlist "${playlist.name}"?`)) {
        await playlistStore.delete(playlist)
        toastSuccess(`Playlist "${playlist.name}" deleted.`)
        router.go('home')
      }
    }).on('PLAYLIST_FOLDER_DELETE', async folder => {
      if (await showConfirmDialog(`Delete the playlist folder "${folder.name}"?`)) {
        await playlistFolderStore.delete(folder)
        toastSuccess(`Playlist folder "${folder.name}" deleted.`)
        router.go('home')
      }
    }).on('LOG_OUT', async () => {
      await userStore.logout()
      authService.destroy()
      forceReloadWindow()
    })

    return () => slots.default?.()
  }
})
