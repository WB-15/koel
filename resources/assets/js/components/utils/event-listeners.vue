<template>
  <span></span>
</template>
<script lang="ts" setup>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { auth } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'

eventBus.on({
  PLAYLIST_DELETE (playlist: Playlist) {
    const destroy = async () => {
      await playlistStore.delete(playlist)
      alerts.success(`Deleted playlist "${playlist.name}."`)
      router.go('home')
    }

    if (!playlist.songs.length) {
      destroy()
    } else {
      alerts.confirm(`Delete the playlist "${playlist.name}"?`, destroy)
    }
  },

  /**
   * Log the current user out and reset the application state.
   */
  async LOG_OUT () {
    await userStore.logout()
    auth.destroy()
    forceReloadWindow()
  },

  'KOEL_READY': () => router.init(),

  /**
   * Hide the panel away if a main view is triggered on mobile.
   */
  'LOAD_MAIN_CONTENT': () => isMobile.phone && (preferenceStore.showExtraPanel = false)
})
</script>
