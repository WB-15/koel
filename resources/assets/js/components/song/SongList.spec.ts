import { expect, it } from 'vitest'
import { ref } from 'vue'
import factory from '@/__tests__/factory'
import UnitTestCase from '@/__tests__/UnitTestCase'
import { arrayify } from '@/utils'
import {
  SelectedSongsKey,
  SongListConfigKey,
  SongListSortFieldKey,
  SongListSortOrderKey,
  SongListTypeKey,
  SongsKey
} from '@/symbols'
import SongList from './SongList.vue'
import { fireEvent } from '@testing-library/vue'

let songs: Song[]

new class extends UnitTestCase {
  private renderComponent (
    _songs: Song | Song[],
    type: SongListType = 'all-songs',
    config: Partial<SongListConfig> = {},
    selectedSongs: Song[] = [],
    sortField: SongListSortField = 'title',
    sortOrder: SortOrder = 'asc'
  ) {
    songs = arrayify(_songs)

    return this.render(SongList, {
      global: {
        stubs: {
          VirtualScroller: this.stub('virtual-scroller')
        },
        provide: {
          [SongsKey]: [ref(songs)],
          [SelectedSongsKey]: [ref(selectedSongs), value => selectedSongs = value],
          [SongListTypeKey]: [ref(type)],
          [SongListConfigKey]: [config],
          [SongListSortFieldKey]: [ref(sortField), value => sortField = value],
          [SongListSortOrderKey]: [ref(sortOrder), value => sortOrder = value]
        }
      }
    })
  }

  protected test () {
    it('renders', async () => {
      const { html } = this.renderComponent(factory<Song[]>('song', 5))
      expect(html()).toMatchSnapshot()
    })

    it.each([
      ['track', 'header-track-number'],
      ['title', 'header-title'],
      ['album_name', 'header-album'],
      ['length', 'header-length'],
      ['artist_name', 'header-artist']
    ])('sorts by %s upon %s clicked', async (field: SongListSortField, testId: string) => {
      const { getByTestId, emitted } = this.renderComponent(factory<Song[]>('song', 5))

      await fireEvent.click(getByTestId(testId))
      expect(emitted().sort[0]).toBeTruthy([field, 'asc'])

      await fireEvent.click(getByTestId(testId))
      expect(emitted().sort[0]).toBeTruthy([field, 'desc'])
    })
  }
}
