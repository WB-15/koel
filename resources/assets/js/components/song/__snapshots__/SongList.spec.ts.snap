// Vitest Snapshot v1

exports[`renders 1`] = `
<div class="all-songs song-list-wrap main-scroll-wrap" data-testid="song-list" tabindex="0">
  <div class="sortable song-list-header"><span class="track-number" data-testid="header-track-number" role="button" title="Sort by track number"> # <!--v-if--><!--v-if--></span><span class="title" data-testid="header-title" role="button" title="Sort by title"> Title <svg class="svg-inline--fa fa-angle-down text-highlight" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="angle-down" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><path class="" fill="currentColor" d="M192 384c-8.188 0-16.38-3.125-22.62-9.375l-160-160c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L192 306.8l137.4-137.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-160 160C208.4 380.9 200.2 384 192 384z"></path></svg><!--v-if--></span><span class="artist" data-testid="header-artist" role="button" title="Sort by artist"> Artist <!--v-if--><!--v-if--></span><span class="album" data-testid="header-album" role="button" title="Sort by album"> Album <!--v-if--><!--v-if--></span><span class="time" data-testid="header-length" role="button" title="Sort by song duration"><!--v-if--><!--v-if--> &nbsp; <svg class="svg-inline--fa fa-clock duration-header" aria-hidden="true" focusable="false" data-prefix="far" data-icon="clock" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path class="" fill="currentColor" d="M232 120C232 106.7 242.7 96 256 96C269.3 96 280 106.7 280 120V243.2L365.3 300C376.3 307.4 379.3 322.3 371.1 333.3C364.6 344.3 349.7 347.3 338.7 339.1L242.7 275.1C236 271.5 232 264 232 255.1L232 120zM256 0C397.4 0 512 114.6 512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256C0 114.6 114.6 0 256 0zM48 256C48 370.9 141.1 464 256 464C370.9 464 464 370.9 464 256C464 141.1 370.9 48 256 48C141.1 48 48 141.1 48 256z"></path></svg></span><span class="favorite"></span><span class="play"></span></div><br data-testid="virtual-scroller" item-height="35" items="">
</div>
`;
